### ISSUE TITLE
Loading strategies

### ISSUE THREAD

**Issue by SnarkBoojum** [fplll/fpylll#221]
After a ```from fpylll import *```,  ```load_strategies_json(BKZ.DEFAULT_STRATEGY)``` fails stating "Cannot open strategies file." ; because in fact it should be: ```load_strategies_json(BKZ.DEFAULT_STRATEGY_PATH + BKZ.DEFAULT_STRATEGY)``` or some such.

Notice that all of the comments/tests in the source point to the broken way ; only check_delta in src/fpylll/fplll/bkz_param.pyx tries to do the right thing by using strategy_full_path.

Arguably fplll's load_strategies_json should detect a relative filename and try to prepend its default path -- that would simplify matters.

**Comment by SnarkBoojum** [comment]
In sagemath, sage/src/sage/matrix/matrix_integer_dense.pyx, the line ```kwds["strategies"] = load_strategies_json(BKZ.DEFAULT_STRATEGY)``` is hence broken.

**Comment by SnarkBoojum** [comment]
Sigh. Since last release, I see it's now even worse! Commit a249d630debd8e96d536c37762f250d562aaec65 makes fpylll search for SAGE!? It's definitely an incorrect change: sagemath's fplll package should be configured so its DEFAULT_STRATEGY_PATH points to where default.json is installed, and fpylll should use just that!

**Comment by dimpase** [comment]
Indeed, there should be no fooling around with Sage/`SAGE_LOCAL` here,  the strategies info is in `/usr/include/fplll/fplll_config.h`: see macros
`FPLLL_DEFAULT_STRATEGY_PATH` and `FPLLL_DEFAULT_STRATEGY`
so it's a matter for a simple (say) Cython function to get these value for fpylll.  Although it could be even better to store them in `fplll.pc`- then it could be obtained from Python very easily.

**Comment by dimpase** [comment]
See #133

**Comment by dimpase** [comment]
@joerowell

**Comment by SnarkBoojum** [comment]
In fact, I think the best fix is a single-liner in fplll's load_strategies_json:
```std::ifstream fs(filename);``` in its source should become ```std::ifstream fs(strategy_full_path(filename));```

**Comment by SnarkBoojum** [comment]
Reported on fplll [here](https://github.com/fplll/fplll/issues/485).

This issue on fpylll is now about purging the code from anything fancy about sagemath.

**Comment by joerowell** [comment]
Happy to purge the Sagemath stuff from here: the suggested idea seems much better (mea culpa!) 

I can't replicate the bug on my Debian box, but I'll trust from the reports that it exists and go from there.

**Comment by dimpase** [comment]
to fix this in Sage, I opened https://trac.sagemath.org/ticket/33302

**Comment by dimpase** [comment]
> In sagemath, sage/src/sage/matrix/matrix_integer_dense.pyx, the line `kwds["strategies"] = load_strategies_json(BKZ.DEFAULT_STRATEGY)` is hence broken.

I think this is probably the best place to fix this. The biggest problem here is the lack of documentation on fp(y)lll side - how are these values,  BKZ.DEFAULT_STRATEGY and BKZ.DEFAULT_STRATEGY_PATH, formed?
Are they just take what's in the fplll header (macros DEFAULT_STRATEGY and DEFAULT_STRATEGY_PATH), and the latter
are always correct (since some version of fplll)? Or it's something more complicated?

The fix suggested by @SnarkBoojum might be the best for Debian, but it's suboptimal for mutiplatform Python libaries fpylll and Sage, which might rely on pre-compiled fplll, IMHO.

**Comment by dimpase** [comment]
It's all quite messy. While talking about Debian 11's fplll, it's version 5.4.0, with FPLLL_DEFAULT_STRATEGY and FPLLL_DEFAULT_STRATEGY_PATH set to the correct path, and to `default.json`, respectively, in `fplll/fplll_config.h`. But on
e.g. Fedora 34, there is fplll 5.4.1, and there are no such macros in `fplll/fplll_config.h`. Instead, these macros as in `fplll/defs.h`, and they are set to `""`. Somehow, on Fedora 34 in Sage 9.5,  `BKZ.DEFAULT_STRATEGY` is getting set to the correct full path to `default.json` (in `/usr/share/fplll/strategies`).

How this is happening, is hard to say without knowing the corresponding change in fplll in 5.4.1.

**Comment by dimpase** [comment]
I'm trying to compile and run (with `g++ foo.C -lfplll && ./a.out`)
```C++
/* foo.C */
#include <iostream>
#include <fplll/fplll.h>
#include <fplll/bkz_param.h>
int main ()
{
std::cout << fplll::default_strategy_path() << "\n";
return 0;
}
```
on various systems. So on Fedora 34 (with unpatched (?) fplll 5.4.1) I get `/usr/share/fplll/strategies`, as it should be, even
though this value is not in the headers anywhere. So it seems to be hardcoded during build time.

On Debian 11, with fplll 5.4.0, I get `/usr/share/libfplll7/strategies/`, as it should be, and this same value is in the headers.

**Comment by SnarkBoojum** [comment]
It's actually handled in fplll's Makefile.am, and I added [a patch](https://salsa.debian.org/science-team/fplll/-/blob/master/debian/patches/versioned_strategies_path) in Debian so it's versioned (make libfplll257 libfplll314 coinstallable). In 2017 according to git blame.

**Comment by dimpase** [comment]
It seems to be a good patch, why isn't it upstreamed? The current situation with the headers in 5.4.1 is confusing at best.

**Comment by SnarkBoojum** [comment]
Well, I  forwarded it to upstream but it wasn't applied... perhaps you can propose again?

**Comment by dimpase** [comment]
> libfplll6 and libfplll7 are co-installable, and each can find its correct strategies file.

yes, but FPLLL_DEFAULT_STRATEGY* macros will only be correct for one of them.

**Comment by SnarkBoojum** [comment]
Those macros set the path at compile-time, while libfplll<n> is a runtime dep. Link to the right one, run with the right one, find the right files. No problem.

**Comment by dimpase** [comment]
why are these macros needed in the header file at all?

**Comment by joerowell** [comment]
(Just to make sure I understand the state of this issue) is your argument:

- that the macros could be placed elsewhere during building, or

- that we shouldn't have the macros at all, and instead store this
information differently?

Is there anything extra that needs to be done _here_ for this? From what I
can tell this looks like a building issue, which I'm inclined to look into
as a fplll issue (rather than an fpylll issue).

**Comment by SnarkBoojum** [comment]
If fplll's api only allowed "load_strategies()", they wouldn't be needed.

But its "load_strategies(path + filename)", so the downstream developers need to access the default path and default filename to actually call that function with meaningful values.

**Comment by dimpase** [comment]
From C++ it's all available via concatenating `fplll::default_strategy_path()` and `fplll::default_strategy()`.
The values are baked in at build time.

**Comment by joerowell** [comment]
The patch looks fine, but it's not entirely satisfying for fpylll (if there's an issue with paths in fplll we should fix it across all systems).

@malb what do you think? I'm thinking of #133 primarily, but also the examples that @dimpase has brought up for e.g Fedora.

**Comment by dimpase** [comment]
> is there anything else we should do?

as long as you don't break what you have now regarding `BKZ.DEFAULT_STRATEGY_PATH` and `BKZ.DEFAULT_STRATEGY`,
all is good and taken care of by the fix in Sage ticket.

**Comment by SnarkBoojum** [comment]
I wanted to upload the latest fpylll to Debian, but had to add this patch, because it insisted on looking at the wrong place:
```
--- fpylll.orig/src/fpylll/fplll/bkz_param.pyx
+++ fpylll/src/fpylll/fplll/bkz_param.pyx
@@ -26,7 +26,7 @@
 from fpylll.util cimport check_delta, check_pruner_metric
 from cython.operator cimport dereference as deref, preincrement as inc
 
-from fpylll.config import default_strategy
+from fpylll.config import default_strategy_path, default_strategy
 
 from .pruner cimport PruningParams
 
@@ -278,8 +278,8 @@
     #  catch. If it just so happens that we fail, we read the Sagemath environment variables and try
     #  to navigate to the default file relative to that directory.
 
-    if not path.exists(filename) and filename == "default.json":
-        filename = path.join(environ['SAGE_LOCAL'], 'fplll', 'strategies', 'default.json')
+    if not path.exists(filename):
+        filename = path.join(default_strategy_path, filename)
 
     if not path.exists(filename):
         raise FileNotFoundError("File '%s' not found."%filename)
```

**Comment by dimpase** [comment]
handling of  `default.json` was seriously messed up, by too eager changes done by fpylll devs to fix upstream a Sage issue. I recall untangling it a while ago.

**Comment by dimpase** [comment]
of course it's an fpylll bug to assume that `SAGE_LOCAL` is always set.

**Comment by SnarkBoojum** [comment]
Well, it's a fpylll bug to even look at SAGE_LOCAL!

It should really only care bout default_strategy_path, and sagemath should compile its fplll with the right path.

**Comment by dimpase** [comment]
> Well, it's a fpylll bug to even look at SAGE_LOCAL!
> 
> It should really only care bout default_strategy_path, and sagemath should compile its fplll with the right path.

indeed. I have no idea why they decided to mess up with `SAGE_LOCAL` in the first place.

**Comment by dimpase** [comment]
I proposed to write the location of `default.json` into a variable in `fplll.pc`, say `strategypath`.

Then it can be inspected from shell, or from python, or from autoconf, without any trouble. E.g from shell
```shell
pkg-config --variable=strategypath fplll
```

**Comment by SnarkBoojum** [comment]
I don't really care how it's available, as long as fplll can be configured to lay its hand on it and fpylll doesn't break everything in its back...

**Comment by SnarkBoojum** [comment]
Notice that Debian's fplll uses a patch as mentioned [here](https://github.com/fplll/fplll/issues/320).

**Comment by malb** [comment]
Oh wow, you two are angry! :) I'll do a PR to clean this up, can I then ask you to review it (too)?

**Comment by dimpase** [comment]
> Notice that Debian's fplll uses a patch as mentioned [here](https://github.com/fplll/fplll/issues/320).

As far as I know, `default.strategy` is installed by some kind of autotools magic. I'm surprised that it does not honour `--prefix` - or perhaps it does, but not in the way Debian wants it, no idea.

**Comment by dimpase** [comment]
> Oh wow, you two are angry! :) I'll do a PR to clean this up, can I then ask you to review it (too)?

Yeah, IMHO `fpylll.config` is due for cleaning, and IMHO should provide a ready to use way to get `default.strategy`. It's a part of configuration, after all, no?

**Comment by SnarkBoojum** [comment]
The pristine 0.5.8 testsuite breaks on Debian because default strategies aren't found.

If a distribution packages fplll correctly, then fpylll should just work ; adding workarounds for a specific distribution and breaking the others is not a good software engineering practice.

Here is the patch I'm applying to package 0.5.8:
```
Description: unbreak finding the default strategy
Author: Julien Puydt
Forwarded: https://github.com/fplll/fpylll/issues/221

--- fpylll.orig/src/fpylll/fplll/bkz_param.pyx
+++ fpylll/src/fpylll/fplll/bkz_param.pyx
@@ -263,16 +263,7 @@
 
     """
 
-    #  It turns out that on some platforms (e.g pre-compiled Sagemath) the BKZ.DEFAULT_STRATEGY
-    #  path gets a bit broken. If this happens there's a C++ exception thrown from fplll, which we
-    #  catch. If it just so happens that we fail, we read the Sagemath environment variables and try
-    #  to navigate to the default file relative to that directory.
-
-    if not path.exists(filename) and filename == "default.json":
-        if 'SAGE_LOCAL' in environ:
-            # NOTE: some people are quite angry about this: https://github.com/fplll/fpylll/issues/221
-            filename = path.join(environ['SAGE_LOCAL'], 'fplll', 'strategies', 'default.json')
-        else:
+    if not path.exists(filename):
             filename = path.join(default_strategy_path, filename)
 
     if not path.exists(filename):
```

**Comment by malb** [comment]
So fundamentally, you're dropping the ` and filename == "default.json"` check. What is the value when it breaks?

**Comment by SnarkBoojum** [comment]
Well, fundamentally I'm simplifying the code to rely on fplll's packaging not being broken... and since I'm also a fplll packager, it's not!

The value was "default.json".

**Comment by malb** [comment]
I'm struggling to see how your code would behave differently on that input than what's there. What am I missing?

**Comment by SnarkBoojum** [comment]
I don't know ; the current code wasn't working, the new one does.

Too smart, broken: KISS!

**Comment by musicinmybrain** [comment]
Hi, I’m the current primary maintainer of the [`libfplll` package in Fedora Linux](https://src.fedoraproject.org/rpms/libfplll). I just learned that the `FPLLL_DEFAULT_STRATEGY` and `FPLLL_DEFAULT_STRATEGY_PATH` macros in the headers don’t have useful values. That’s obviously a bug, and I’m happy to fix it in Fedora one way or another.

It looks like `fplll` upstream doesn’t have any provision to define those properly when building packages; the empty-string fallback values are defined in `fplll/defs.h`, and the macros aren’t otherwise defined in any headers.

https://github.com/fplll/fplll/blob/3e10c51b31f29fca4a6b1e11494b96a5009dd19d/fplll/defs.h#L283

The reason that the experiment in https://github.com/fplll/fpylll/issues/221#issuecomment-1031488375 yielded appropriate values is that these macros are defined using compiler flags while building the library:

https://github.com/fplll/fplll/blob/3e10c51b31f29fca4a6b1e11494b96a5009dd19d/fplll/Makefile.am#L5

so they are available while compiling [`fplll/bkz_param.cpp`](https://github.com/fplll/fplll/blob/5.4.4/fplll/bkz_param.cpp) and are thus compiled into the library.

It seems like the default values used at build time should be formed in `configure.ac` and `AC_SUBST`ed into [`fplll/fplll_config.h.in`](https://github.com/fplll/fplll/blob/5.4.4/fplll/fplll_config.h.in), which produces `fplll/fplll_config.h`, which is an API header and is `#include`d from `fplll/defs.h`. That would make the defaults used at compile time available via the API headers, just as is already the case for some compile-time values like `FPLLL_MAX_ENUM_DIM`. I’ll propose that to `fplll` and see if they are interested in a PR.

**Comment by musicinmybrain** [comment]
Comments welcome over in https://github.com/fplll/fplll/issues/510, where I’m trying to put together an approach that will make this workable for everyone.

**Comment by musicinmybrain** [comment]
All interested parties, please review https://github.com/fplll/fplll/pull/512 and let me know if it meets your needs.

**Comment by musicinmybrain** [comment]
> All interested parties, please review [fplll/fplll#512](https://github.com/fplll/fplll/pull/512) and let me know if it meets your needs.

Polite ping; I haven’t seen any feedback on this PR. Thanks!
